tests/unit/drivers/test_recursive_traversal.py

```diff
@@ -198,3 +198,43 @@ def test_adjacency1_granularity1():
     assert len(docs[0].matches[0].matches) == DOCUMENTS_PER_LEVEL
     assert len(docs[0].matches[0].matches[0].chunks) == DOCUMENTS_PER_LEVEL
 
+
+def test_adjacency_on_chunks():
+    docs = build_docs()
+    doc = docs[0]
+    doc.ClearField('matches')
+    docs = [doc, ]
+    driver = SliceQL(
+        start=0,
+        end=1,
+        adjacency_range=(1, 1),
+        granularity_range=(1, 2),
+        recur_on=["chunks", ]
+    )
+    driver._traverse_apply(docs)
+    assert len(docs) == 1
+    assert len(docs[0].chunks) == 1
```

Comment:
This should not be the case. `adjacency_range=(1,1)` says, that the chunks should only be touched beginning from the first adjacency layer and not from the original document.

Reply:
I'm a bit confused. This means the setting of `adjacency_range` is coupled with `granularity_range` if though I just want to run either on `chunks` or `matches`.

Reply:
By this setting, my expectation is that it works on `chunks` from `granularity_range=1` to `2`.

Reply:
But on which `adjacency_depth`? So do you want to touch the chunks of the document with `adjacency_depth=0` or the chunks or the documents with `adjacency_depth=1`? Since we want to provide a possibility to iterate through the chunks of the documents with `adjacency_depth=1`, setting `adjacency_range=(1,1)` is the way to do that. If you now set `adjacency_range=(1,1)` and the chunks of the document with `adjacency_depth=0` are touched, that is confusing.

I think making this less confusing might be in decoupling the parameters further down (with example values):
- `recur_on=['chunks', 'matches']` same meaning as it has now
- `chunks_iteration_adjacency_start_depth=1` start iterating over chunks on the `adjacency_depth=1` and `granularity_depth=0` along the `chunks` edges (only has an effect if `chunks` is in `recur_on`)
- `granularity_range=(0,1)` touch all chunks on the `granularity_depth` zero and one (only has an effect if `chunks` is in `recur_on`)
- `matches_iteration_granularity_start_depth=2` start iterating over matches on the `granularity_depth=2` and `adjacency_depth=0` along the `matches` edges (only has an effect if `matches` is in `recur_on`)
- `adjacency_range=(1,2)` touch all matches on the `adjacency_depth` one and two (only has an effect if `matches` is in `recur_on`)

Beware the difference between iterating and touch in this description: 
- iterating: traverse through the document
- touch: call the `_apply` or `_apply_all` functions on the document

Reply:
When you now take this variables and set `chunks_iteration_adjacency_start_depth=1`, even thought the root document has no matches, what should happen? I think it should (gracefully?) fail.